File: quak/quack.py
import requests
import time


def get_balance(token):
    balance_url = 'https://api.quackquack.games/balance/get'
    headers = {
    'Accept': '*/*',
    'Accept-Encoding': 'gzip, deflate, br, zstd',
    'Accept-Language': 'en-US,en;q=0.9,vi;q=0.8,zh-CN;q=0.7,zh;q=0.6',
    'Authorization': f'Bearer {token}',
    'Content-Type': 'application/x-www-form-urlencoded',
    'Origin': 'https://dd42189ft3pck.cloudfront.net',
    'Referer': 'https://dd42189ft3pck.cloudfront.net/',
    'Sec-Ch-Ua': '"Chromium";v="124", "Google Chrome";v="124", "Not-A.Brand";v="99"',
    'Sec-Ch-Ua-Mobile': '?0',
    'Sec-Ch-Ua-Platform': '"Windows"',
    'Sec-Fetch-Dest': 'empty',
    'Sec-Fetch-Mode': 'cors',
    'Sec-Fetch-Site': 'cross-site',
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
}
    response = requests.get(balance_url, headers=headers)
    if response.status_code == 200:
        balances = response.json().get('data', {}).get('data', [])
        egg_balance_info = next((item for item in balances if item['symbol'] == 'EGG'), None)
        if egg_balance_info:
            return egg_balance_info['balance']
        else:
            return "Không tìm thấy thông tin balance của EGG."
    else:
        return "Lỗi khi lấy balance."
def get_nest_ids(token):
    list_url = 'https://api.quackquack.games/nest/list-reload'
    headers = {
    'Accept': '*/*',
    'Accept-Encoding': 'gzip, deflate, br, zstd',
    'Accept-Language': 'en-US,en;q=0.9,vi;q=0.8,zh-CN;q=0.7,zh;q=0.6',
    'Authorization': f'Bearer {token}',
    'Content-Type': 'application/x-www-form-urlencoded',
    'Origin': 'https://dd42189ft3pck.cloudfront.net',
    'Referer': 'https://dd42189ft3pck.cloudfront.net/',
    'Sec-Ch-Ua': '"Chromium";v="124", "Google Chrome";v="124", "Not-A.Brand";v="99"',
    'Sec-Ch-Ua-Mobile': '?0',
    'Sec-Ch-Ua-Platform': '"Windows"',
    'Sec-Fetch-Dest': 'empty',
    'Sec-Fetch-Mode': 'cors',
    'Sec-Fetch-Site': 'cross-site',
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
}

    response = requests.get(list_url, headers=headers)
    if response.status_code == 200:
        nest_data = response.json().get('data', {}).get('nest', [])
        return [(nest['id'], nest.get('type_egg')) for nest in nest_data]
    else:
        print("Không thể lấy danh sách nest. Mã lỗi:", response.status_code)
        return []

def collect_nest(nest_id, token):
    collect_url = 'https://api.quackquack.games/nest/collect'
    headers = {
        'Accept': '*/*',
        'Accept-Encoding': 'gzip, deflate, br, zstd',
        'Accept-Language': 'en-US,en;q=0.9,vi;q=0.8,zh-CN;q=0.7,zh;q=0.6',
        'Authorization': f'Bearer {token}',
        'Content-Type': 'application/x-www-form-urlencoded',
        'Origin': 'https://dd42189ft3pck.cloudfront.net',
        'Referer': 'https://dd42189ft3pck.cloudfront.net/',
        'Sec-Ch-Ua': '"Chromium";v="124", "Google Chrome";v="124", "Not-A.Brand";v="99"',
        'Sec-Ch-Ua-Mobile': '?0',
        'Sec-Ch-Ua-Platform': '"Windows"',
        'Sec-Fetch-Dest': 'empty',
        'Sec-Fetch-Mode': 'cors',
        'Sec-Fetch-Site': 'cross-site',
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
    }
    data = {'nest_id': nest_id}
    response = requests.post(collect_url, headers=headers, data=data)
    return response

token = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOjg2NDExOSwidGltZXN0YW1wIjoxNzE4MTcyODM0MDE4LCJ0eXBlIjoxLCJpYXQiOjE3MTgxNzI4MzQsImV4cCI6MTcxODc3NzYzNH0.3_q6x8eQxiygeejBJUxUYjTWCgCUdFlyGHqfoXTMWDA'

while True:
    nest_ids = get_nest_ids(token)

    for nest_id, type_egg in nest_ids:
        if type_egg is None:
            continue
        response = collect_nest(nest_id, token)
        response_data = response.json()  
        balance = get_balance(token)

        if response.status_code == 200 and response_data.get("error_code") == "":
            result_message = "Thành công"
        elif response_data.get("error_code") == "THIS_NEST_DONT_HAVE_EGG_AVAILABLE":
            result_message = "Thất bại - Tổ này không có trứng"
        else:
            result_message = "Lỗi không xác định"

        print(f"Thu hoạch ổ trứng {nest_id}: Status {response.status_code}, Kết quả: {result_message}, EGG balance: {balance}")

        time.sleep(5)
#nghỉ 10s mỗi lượt
time.sleep(10)

File: quak/quack.js
const ACCESS_TOKEN = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOjg2NDExOSwidGltZXN0YW1wIjoxNzE4MzQzNDQyMzE0LCJ0eXBlIjoxLCJpYXQiOjE3MTgzNDM0NDIsImV4cCI6MTcxODk0ODI0Mn0.-kHnbdADPe9NQA0KOwtJzCI-NVIaDG-6e64HmaxuPUw";
let listColect = [];
let listDuck = [];

Array.prototype.random = function () {
  return this[Math.floor(Math.random() * this.length)];
};

function getTotalEgg() {
  fetch("https://api.quackquack.games/balance/get", {
    headers: {
      accept: "*/*",
      "accept-language": "en-US,en;q=0.9,vi;q=0.8",
      authorization: "Bearer " + ACCESS_TOKEN,
      "if-none-match": 'W/"1a9-I7Onn3jBU9AHo0MlzSY8mMECNvQ"',
      priority: "u=1, i",
      "sec-ch-ua":
        '"Google Chrome";v="125", "Chromium";v="125", "Not.A/Brand";v="24"',
      "sec-ch-ua-mobile": "?0",
      "sec-ch-ua-platform": '"Windows"',
      "sec-fetch-dest": "empty",
      "sec-fetch-mode": "cors",
      "sec-fetch-site": "cross-site",
      Referer: "https://dd42189ft3pck.cloudfront.net/",
      "Referrer-Policy": "strict-origin-when-cross-origin",
    },
    body: null,
    method: "GET",
  })
    .then((response) => response.json())
    .then((res) => {
      // console.log(res);
      res.data.data.map((item) => {
        if (item.symbol === "EGG") {
          console.log("");
          console.log("Da thu thap", Number(item.balance), "trung");
          console.log("");
        }
      });
    })
    .catch((error) => {
      console.log("ERROR", error);
    });
}

function getListReload() {
  fetch("https://api.quackquack.games/nest/list-reload", {
    headers: {
      accept: "*/*",
      "accept-language": "en-US,en;q=0.9,vi;q=0.8",
      authorization: "Bearer " + ACCESS_TOKEN,
      "if-none-match": 'W/"1218-LZvWPzXbQkzjfWJ5mauEo0z3f9c"',
      priority: "u=1, i",
      "sec-ch-ua":
        '"Google Chrome";v="125", "Chromium";v="125", "Not.A/Brand";v="24"',
      "sec-ch-ua-mobile": "?0",
      "sec-ch-ua-platform": '"Windows"',
      "sec-fetch-dest": "empty",
      "sec-fetch-mode": "cors",
      "sec-fetch-site": "cross-site",
      Referer: "https://dd42189ft3pck.cloudfront.net/",
      "Referrer-Policy": "strict-origin-when-cross-origin",
    },
    body: null,
    method: "GET",
  })
    .then((response) => response.json())
    .then((res) => {
      // console.log(res);

      if (listDuck.length === 0) {
        res.data.duck.map((item) => {
          listDuck.push(item.id);
        });
      }

      res.data.nest.map((item) => {
        // console.log(item);
        if (item.type_egg) listColect.push(item.id);
      });

      console.log("So trung co the thu thap:", listColect.length);
      console.log(listColect);
      console.log("");
    })
    .catch((error) => {
      console.log("ERROR", error);
    });
}

function collect() {
  if (listColect.length === 0) return setTimeout(collect, 3e3);

  const egg = listColect[0];

  fetch("https://api.quackquack.games/nest/collect", {
    headers: {
      accept: "*/*",
      "accept-language": "en-US,en;q=0.9,vi;q=0.8",
      authorization: "Bearer " + ACCESS_TOKEN,
      "content-type": "application/x-www-form-urlencoded",
      priority: "u=1, i",
      "sec-ch-ua":
        '"Google Chrome";v="125", "Chromium";v="125", "Not.A/Brand";v="24"',
      "sec-ch-ua-mobile": "?0",
      "sec-ch-ua-platform": '"Windows"',
      "sec-fetch-dest": "empty",
      "sec-fetch-mode": "cors",
      "sec-fetch-site": "cross-site",
      Referer: "https://dd42189ft3pck.cloudfront.net/",
      "Referrer-Policy": "strict-origin-when-cross-origin",
    },
    body: "nest_id=" + egg,
    method: "POST",
  })
    .then((response) => response.json())
    .then((res) => {
      //   console.log(res);
      console.log("Thu thap thanh cong trung", egg);
      layEgg(egg);
    })
    .catch((error) => {
      console.log("ERROR", error);
      setTimeout(() => {
        collect(egg);
      }, 3e3);
    });
}

function layEgg(egg) {
  const duck = listDuck.random();
  //   console.log(duck);
  fetch("https://api.quackquack.games/nest/lay-egg", {
    headers: {
      accept: "*/*",
      "accept-language": "en-US,en;q=0.9,vi;q=0.8",
      authorization: "Bearer " + ACCESS_TOKEN,
      "content-type": "application/x-www-form-urlencoded",
      priority: "u=1, i",
      "sec-ch-ua":
        '"Google Chrome";v="125", "Chromium";v="125", "Not.A/Brand";v="24"',
      "sec-ch-ua-mobile": "?0",
      "sec-ch-ua-platform": '"Windows"',
      "sec-fetch-dest": "empty",
      "sec-fetch-mode": "cors",
      "sec-fetch-site": "cross-site",
      Referer: "https://dd42189ft3pck.cloudfront.net/",
      "Referrer-Policy": "strict-origin-when-cross-origin",
    },
    body: "nest_id=" + egg + "&duck_id=" + duck,
    method: "POST",
  })
    .then((response) => response.json())
    .then((res) => {
      //   console.log(res);
      getTotalEgg();
      listColect.shift();
      setTimeout(collect, 3e3);
    })
    .catch((error) => {
      console.log("ERROR", error);
      setTimeout(() => {
        layEgg(egg);
      }, 3e3);
    });
}

getTotalEgg();
getListReload();

setInterval(() => {
  getListReload();
}, 10e3);

collect();